samples/guide/src/main/java/okhttp3/recipes/UploadProgress.java

```diff
@@ -0,0 +1,134 @@
+
+public final class UploadProgress {
+	private static final String IMGUR_CLIENT_ID = "9199fdef135c122";
+	private static final MediaType MEDIA_TYPE_PNG = MediaType.get("image/png");
+
+	private final OkHttpClient client = new OkHttpClient();
+
+	public void run() throws Exception {
+		// Use the imgur image upload API as documented at https://api.imgur.com/endpoints/image
+		final ProgressListener progressListener = new ProgressListener() {
+			boolean firstUpdate = true;
+
+			@Override public void update(long bytesWritten, long contentLength, boolean done) {
+				if (done) {
+					System.out.println("completed");
+				} else {
+					if (firstUpdate) {
+						firstUpdate = false;
+						if (contentLength == -1) {
+							System.out.println("content-length: unknown");
+						} else {
+							System.out.format("content-length: %d\n", contentLength);
+						}
+					}
+
+					System.out.println(bytesWritten);
+
+					if (contentLength != -1) {
+						System.out.format("%d%% done\n", (100 * bytesWritten) / contentLength);
+					}
+				}
+			}
+		};
+
+		RequestBody requestBody = RequestBody.create(
+			new File("docs/images/logo-square.png"),
+			MEDIA_TYPE_PNG);
+
+		Request request = new Request.Builder()
+			.header("Authorization", "Client-ID " + IMGUR_CLIENT_ID)
+			.url("https://api.imgur.com/3/image")
+			.post(new ProgressRequestBody(requestBody, progressListener))
+			.build();
+
+		Response response = client.newCall(request).execute();
+		if (!response.isSuccessful()) throw new IOException("Unexpected code " + response);
+
+		System.out.println(response.body().string());
+	}
+
+	public static void main(String... args) throws Exception {
+		new UploadProgress().run();
+	}
+
+	private static class ProgressRequestBody extends RequestBody {
+
+		private final ProgressListener progressListener;
+		private final RequestBody requestBody;
+
+		public ProgressRequestBody(RequestBody requestBody, ProgressListener progressListener) {
+			this.requestBody = requestBody;
+			this.progressListener = progressListener;
+		}
+
+		@Override
+		public MediaType contentType() {
+			return requestBody.contentType();
+		}
+
+		@Override
+		public long contentLength() throws IOException {
+			return requestBody.contentLength();
+		}
+
+		@Override
+		public void writeTo(BufferedSink sink) throws IOException {
+			BufferedSink bufferSink = Okio.buffer(sink(sink));
+			requestBody.writeTo(bufferSink);
+			bufferSink.close();
+		}
+
+		public Sink sink(Sink sink) {
+			return new ForwardingSink(sink) {
+				private long totalBytesWritten = 0L;
+
+				@Override
+				public void write(Buffer source, long byteCount) throws IOException {
+					super.write(source, byteCount);
+					totalBytesWritten += byteCount;
+					progressListener.update(totalBytesWritten, contentLength(), false);
+				}
+
+				@Override
+				public void close() throws IOException {
+					super.close();
+					progressListener.update(totalBytesWritten, contentLength(),true);
```

Comment:
supernit: formatting. Space after `,`

Reply:
MIGHT want to make this only call `progressListener.update()` once even if `close()` is called multiple times. You’re allowed to call `close()` multiple times but listeners shouldn’t be called more than once